modules/core/test/src/smithy4s/CachedSchemaVisitorSpec.scala

```diff
@@ -21,156 +21,187 @@ import smithy4s.schema._
-      None
     }
 
     def refine[A, B](
         schema: Schema[A],
         refinement: Refinement[A, B]
-    ): Option[B] = {
+    ): Unit = discard {
       self(schema)
       counter.incrementAndGet()
-      None
     }
 
-    def lazily[A](suspend: Lazy[Schema[A]]): Option[A] = {
-      counter.incrementAndGet()
-      self(suspend.value)
+    def lazily[A](suspend: Lazy[Schema[A]]): Unit = discard {
+      val shapeId = suspend.value.shapeId
+      if (!lazyTracker.contains(suspend.value.shapeId)) {
+        lazyTracker.add(shapeId)
+        counter.incrementAndGet()
+        self(suspend.value)
+        lazyTracker.remove(shapeId)
+      }
     }
   }
 
-  test("Sanity check should NOT hit cache on first visit") {
+  def checkSchema[A](schema: Schema[A]): Unit = {
+    val cache = CompilationCache.make[ConstUnit]
     val counter = new AtomicInteger(0)
-    val visitor = new TestSchemaVisitor(counter)
+    val visitor = new TestSchemaVisitor(counter, cache)
     val schema = Schema.int
-    val result: Option[Int] = visitor(schema)
-    assertEquals(result, None)
-    assertEquals(counter.get(), 1)
+    visitor(schema)
+    val before = counter.get()
+    visitor(schema)
+    val after = counter.get()
+    assertEquals(before, after)
   }
 
-  test(
-    "should hit the cache on second visit with the same schema resulting in counter = 1"
-  ) {
-    val counter = new AtomicInteger(0)
-    val visitor = new TestSchemaVisitor(counter)
-    val schema = Schema.int
-    val result: Option[Int] = visitor(schema)
-    val _ = visitor(schema)
-    assertEquals(result, None)
-    assertEquals(counter.get(), 1)
+  val header = "Cache should prevent recomputation: "
+
+  test(header + "primitive") {
+    checkSchema(int)
   }
 
-  test(
-    "should miss cache on second visit with different schema and the same ShapeId"
-  ) {
-    val counter = new AtomicInteger(0)
-    val visitor = new TestSchemaVisitor(counter)
-    val schema = Schema.unit
-    val lazySchema = Schema.recursive(schema)
-    val result = visitor(schema)
-    val _ = visitor(lazySchema)
-    assertEquals(result, None)
-    assertEquals(counter.get(), 2)
+  test(header + "collection") {
+    checkSchema(list(int))
   }
 
-  test(
-    "test struct with multiple schemas "
-  ) {
-    case class Foo(i: Int, s: String, b: Boolean)
-    val counter = new AtomicInteger(0)
-    val visitor = new TestSchemaVisitor(counter)
-    val schema: Schema[Foo] = struct(
-      int.required[Foo]("i", _.i),
-      string.required[Foo]("s", _.s),
-      boolean.required[Foo]("b", _.b)
-    ) {
-      Foo.apply
+  test(header + "map") {
+    checkSchema(map(string, int))
+  }
+
+  test(header + "enum") {
+    sealed abstract class FooBar(val stringValue: String, val intValue: Int)
+        extends smithy4s.Enumeration.Value {
+      val name = stringValue
+      val value = stringValue
+      val hints = Hints.empty
     }
-    val lazySchema = Schema.recursive(schema)
```

Comment:
Using `Schema.recursive` will not showing any problem if the schema is actually not recursive.